midend/interpreter.cpp

```diff
@@ -310,33 +310,50 @@ void SymbolicStruct::dbprint(std::ostream& out) const {
 }
 
 SymbolicHeaderUnion::SymbolicHeaderUnion(const IR::Type_HeaderUnion* type,
-                               bool uninitialized,
-                               const SymbolicValueFactory* factory) :
-        SymbolicStruct(type, uninitialized, factory),
-        valid(new SymbolicBool(false)) {}
+                                         bool uninitialized,
+                                         const SymbolicValueFactory* factory) :
+        SymbolicStruct(type, uninitialized, factory) {}
+
+void SymbolicHeaderUnion::setValid(bool v, cstring field) {
+    if (field) {
+        fieldValue[field]->to<SymbolicHeader>()->setValid(v);
+        for (auto f : type->to<IR::Type_StructLike>()->fields) {
+            if (f->name.name != field)
+                fieldValue[f->name.name]->setAllUnknown();
+        }
+    }
+}
 
-void SymbolicHeaderUnion::setValid(bool v) {
-    if (!v)
-        setAllUnknown();
-    valid = new SymbolicBool(v);
+SymbolicBool* SymbolicHeaderUnion::isValid() const {
+    for (auto f : type->to<IR::Type_StructLike>()->fields) {
+         auto fieldsClone = fieldValue;
+         if (fieldsClone[f->name.name]) {
+             if (auto fildValid = fieldsClone[f->name.name]->to<SymbolicHeader>()->valid) {
+                 if (fildValid->isKnown())
+                     return new SymbolicBool(fildValid->value);
+             }
+         }
+    }
+    return new SymbolicBool(false);
 }
 
 SymbolicValue* SymbolicHeaderUnion::get(const IR::Node* node, cstring field) const {
-    if (valid->isKnown() && !valid->value)
-        return new SymbolicStaticError(node, "Reading field from invalid header union");
+    if (this->isValid()->isKnown()  && !this->isValid()->value)
+        return new SymbolicStaticError(node,"Reading field from invalid header union");
     return SymbolicStruct::get(node, field);
 }
 
 void SymbolicHeaderUnion::setAllUnknown() {
     SymbolicStruct::setAllUnknown();
-    valid->setAllUnknown();
+    for (auto f : type->checkedTo<IR::Type_StructLike>()->fields) {
```

Comment:
how is this for loop different from the one above?

Reply:
I don't quite understand what you mean

Reply:
setAllUnknown() should do the same thing.

Reply:
I fixed it